pkg/apis/machine/v1beta1/machineset_webhook_test.go

```diff
@@ -0,0 +1,600 @@
+		{
+			name:         "with Azure and no fields set",
+			platformType: osconfigv1.AzurePlatformType,
+			clusterID:    "azure-cluster",
+			providerSpecValue: &runtime.RawExtension{
+				Object: &azure.AzureMachineProviderSpec{},
+			},
+			expectedError: "[providerSpec.location: Required value: location should be set to one of the supported Azure regions, providerSpec.osDisk.diskSizeGB: Invalid value: 0: diskSizeGB must be greater than zero]",
+		},
+		{
+			name:         "with Azure and a location and disk size set",
+			platformType: osconfigv1.AzurePlatformType,
+			clusterID:    "azure-cluster",
+			providerSpecValue: &runtime.RawExtension{
+				Object: &azure.AzureMachineProviderSpec{
+					Location: "location",
+					OSDisk: azure.OSDisk{
+						DiskSizeGB: 128,
+					},
+				},
+			},
+			expectedError: "",
+		},
+		{
+			name:         "with GCP and no fields set",
+			platformType: osconfigv1.GCPPlatformType,
+			clusterID:    "gcp-cluster",
+			providerSpecValue: &runtime.RawExtension{
+				Object: &gcp.GCPMachineProviderSpec{},
+			},
+			expectedError: "providerSpec.region: Required value: region is required",
+		},
+		{
+			name:         "with GCP and the region and zone set",
+			platformType: osconfigv1.GCPPlatformType,
+			clusterID:    "gcp-cluster",
+			providerSpecValue: &runtime.RawExtension{
+				Object: &gcp.GCPMachineProviderSpec{
+					Region: "region",
+					Zone:   "region-zone",
+				},
+			},
+			expectedError: "",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			gs := NewWithT(t)
+
+			mgr, err := manager.New(cfg, manager.Options{
+				MetricsBindAddress: "0",
+				Port:               testEnv.WebhookInstallOptions.LocalServingPort,
+				CertDir:            testEnv.WebhookInstallOptions.LocalServingCertDir,
+			})
+			gs.Expect(err).ToNot(HaveOccurred())
+
+			done := make(chan struct{})
+			go func() {
+				gs.Expect(mgr.Start(done)).To(Succeed())
+			}()
+			defer close(done)
+
+			infra.Status = osconfigv1.InfrastructureStatus{
+				InfrastructureName: tc.clusterID,
+				PlatformStatus: &osconfigv1.PlatformStatus{
+					Type: tc.platformType,
+					GCP: &osconfigv1.GCPPlatformStatus{
+						ProjectID: "gcp-project-id",
+					},
+				},
+			}
+			gs.Expect(c.Status().Update(ctx, infra)).To(Succeed())
+
+			machineSetDefaulter, err := NewMachineSetDefaulter()
+			gs.Expect(err).ToNot(HaveOccurred())
+			machineSetValidator, err := NewMachineSetValidator()
+			gs.Expect(err).ToNot(HaveOccurred())
+			mgr.GetWebhookServer().Register("/mutate-machine-openshift-io-v1beta1-machineset", &webhook.Admission{Handler: machineSetDefaulter})
+			mgr.GetWebhookServer().Register("/validate-machine-openshift-io-v1beta1-machineset", &webhook.Admission{Handler: machineSetValidator})
+
+			ms := &MachineSet{
+				ObjectMeta: metav1.ObjectMeta{
+					GenerateName: "machineset-creation-",
+					Namespace:    namespace.Name,
+				},
+				Spec: MachineSetSpec{
+					Template: MachineTemplateSpec{
```

Comment:
Looking at our API, Template is actually an optional field.  We should ensure that we handle that case and we don't crash.

Reply:
Also, Template.Spec is also optional.

Reply:
While both fields are marked as `omitempty`, as neither is a pointer value, the go type system will infer them even if they are empty, so we don't need to do any checks for nils or anything, I don't think there is anything to handle specifically for these, unless I have missed something?